More detailed errors, Gateway::add_any_port method

canndrew opened the pull request:
This PR parses the XML responses from the gateway and uses them to return more detailed error values. It also adds a new method, `Gateway::add_any_port` which tries to create a port mapping with a random external port. This method internally calls the `AddAnyPortMapping` IGD method, if the method isn't available it calls `add_port` repeatedly with random port values.

canndrew commented:
Also, this PR requires xmltree version 0.3 which isn't yet published. See [here](https://github.com/eminence/xmltree-rs/pull/2).

canndrew commented:
One thing I'm worried about with this PR is that I don't have any way to thoroughly test it. I only have my home router to test on and it doesn't support IGD 2 so I have no way to know if the code that handles a correct `AddAnyPortMapping` response actually works. Also, routers and NAT devices are notorious for being inconsistent and standards non-compliant. My router, for example, returns an invalid/non-standard response to the `DeletePortMapping` method.

I'm wondering if it would be a good idea to try and implement some kind of data collection in this library. The idea is, if the library receives a response from the gateway that it can't make sense of it packages it up with the request that it sent, encrypts it, and emails it to you/me/someone so we can see what problems the library is encountering in the wild. People would have to be warned about this happening with a big, loud warning in the README, it should only happen in debug mode and it should be possible to disable it through a cargo feature. What are your thoughts on this @sbst? Is it a good idea or would it be too much trouble and too much of a privacy violation?

sbst commented:
"What are your thoughts on this @sbst?"
Honestly, I am unable to test your add_any_port method because catch InvalidResponse at gateway.add_port. Probably non-standard response too.
Hence, your idea of collecting and analyzing data is useful. Moreover, it will be better to show some more specific exceptions (more than just "InvalidReponse").

I suppose, sending non-standard messages to your email must be turned off by default (do you really want 10000+ logs on the mail?). In case of some debbuging, user could turn on the sending 
non-standard messages. It warning will be suitable to add in README, of course.
On the other hand, are UPnP answers able to contain some private fields?

P.S. That`s my opinion. Probably, @sbstp has another point of view.

canndrew commented:
> P.S. That`s my opinion. Probably, @sbstp has another point of view.

Lulz sorry, wrong person ;)

sbst commented:
lol
that`s ok

canndrew commented:
Hey, thanks for testing it though! Interesting to know it's broken for your router.

sbstp commented on src/gateway.rs:
```diff
@@ -90,16 +296,182 @@ impl Gateway {
+           local_addr.port(), lease_duration, description);
+
+        let text = try!(soap::send(&url, soap::Action::new(ADD_ANY_PORT_ACTION), &body));
+        println!("got text: {}", text);
```
Forgot debug message.

sbstp commented:
I understand the issue you have, but I don't really want to implement a data collection scheme. It's a bit invasive and adds unnecessary code to the crate. I'd rather state clearly that mileage may vary and that issues should be reported here and will be dealt with. Could probably create a "reporting an issue" guide that asks for dumps of the router's response.

Also you guys at MaidSafe will probably have quite a bit of people using this, so the most common errors, if any, should come out.

sbstp commented:
I just tested on my router, and indeed the `add_any_port` method does not work. The returned string is "Invalid Action" instead of "InvalidAction". I do get an error code though, 401 for an invalid action, it it's in the spec. That might be more reliable.

canndrew commented:
Great, so your router returns "Invalid Action" and mine returns "InvalidAction". I checked the spec and it specifies "Invalid Action" btw. I'll rewrite the PR to use the errorCode rather than errorDescription.

canndrew commented:
Hi, sorry it took a while to get back to this. I ended up making a few changes to the PR:

 * Parse the error codes returned from the gateway rather than the error descriptions.
 * Turn the error types upside-down so `RequestError` is no longer generic. Instead `AddPortError`, `RemovePortError` etc. include a variant to carry a `RequestError`. I think it makes more sense this way.
 * Reorganise the code to make it more concise (mainly abstract some stuff into the `perform_request` method).
 * Changed some error variant names to make them more descriptive rather than using their names from the IGD spec.
 * Added a couple of new error variants.
 * Added an add_any_port example
 * Removed the hard-coded IP address from add_port and make all examples `Display::fmt` errors rather than `Debug::fmt` them.
 * Added `Debug` and `Error` impls for `SearchError`.

I think that's about it. Also this PR is now waiting on [this PR](https://github.com/eminence/xmltree-rs/pull/3). BTW @sbstp I've noticed that you have your own xml tree library, would you prefer that I used that?

vinipsmaker commented:
> Also this PR is now waiting on [this PR](https://github.com/eminence/xmltree-rs/pull/3).

[Not anymore](https://github.com/eminence/xmltree-rs/pull/3#issuecomment-168901700).

sbstp commented:
I'm having issues testing this, the new code seems to be messing with my router. After a few attempts, the router begins to deny every port mapping entry with the 718 code, conflict in mapping entry. I always use a different port, so I'm not too sure what's going on.

sbstp commented:
Never mind this, it seems like I goofed. I think we should change the lease duration to like 60 seconds in the examples, to avoid issues like this. The examples shouldn't leave garbage.

sbstp commented:
Otherwise it looks very good. I think the AddPort/RemovePort/GetExternalIP errors should be exported in lib.rs, since the gateway module is not public. I thought the compiler complained about that stuff.

canndrew commented:
> I think we should change the lease duration to like 60 seconds in the examples
> I think the AddPort/RemovePort/GetExternalIP errors should be exported in lib.rs, since the gateway module is not public.

Done and done.

>  I thought the compiler complained about that stuff.

It didn't, weirdly.

sbstp commented:
Merged, this is available in 0.4.0, it's published on cargo, released on github, and the docs are updated! Thanks a lot.
